### PR TITLE
Add acceptance test and documentation for helm release and feature

### PR DESCRIPTION
**Pull request by Shubbhang351**
1. **What this PR does / why we need it**:

Add acceptance test and documentation for helm release and feature

mock accpe
Screenshot 2023-09-21 at 10 03 26 PM

Screenshot 2023-09-21 at 11 00 23 PM

accpe test
Screenshot 2023-09-22 at 6 14 46 AM

Screenshot 2023-09-22 at 7 02 04 AM

### CODE REVIEW

**Comment by ishangupta-ds** [comment]
@Shubbhang351 please attach a screenshot of the acceptance test run for `helm_feature`.

File: examples/resources/helmfeature/cl_resource.tf [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# Create Tanzu Mission Control cluster scope helm feature with attached set as default value.
+resource "tanzu-mission-control_helm_feature" "create_cl_helm_feature" {
+  scope {
+    cluster {
+      name            = "testcluster" # Required
```
Comment (ramya-bangera) [inline examples/resources/helmfeature/cl_resource.tf:5]: nit: There is terraform formatter target in Makefile , ensure to use it and fix the formatting for all files

File: internal/resources/helmfeature/resource_helm_feature_test.go [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+func TestAcceptanceForHelmFeatureResource(t *testing.T) {
+	testConfig := testGetDefaultAcceptanceConfig(t)
+
+	// If the flag to execute helm feature tests is not found, run this as a mock test by setting up an http intercept for each endpoint.
+	_, found := os.LookupEnv("ENABLE_HELM_ENV_TEST")
+	if !found {
+		os.Setenv("TF_ACC", "true")
+		os.Setenv("TMC_ENDPOINT", "dummy.tmc.mock.vmware.com")
+		os.Setenv("VMW_CLOUD_API_TOKEN", "dummy")
+		os.Setenv("VMW_CLOUD_ENDPOINT", "console.cloud.vmware.com")
```
Comment (ramya-bangera) [inline internal/resources/helmfeature/resource_helm_feature_test.go:60]: org_id is not required here?

File: internal/resources/helmrelease/test_helper.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/vmware/terraform-provider-tanzu-mission-control/internal/authctx"
+	commonscope "github.com/vmware/terraform-provider-tanzu-mission-control/internal/resources/common/scope"
+	helm "github.com/vmware/terraform-provider-tanzu-mission-control/internal/resources/helmfeature"
+)
+
+// nolint: gosec
```
Comment (ramya-bangera) [inline internal/resources/helmrelease/test_helper.go:20]: Which line is throwing lint error here?

**Comment by ramya-bangera** [comment]
LGTM

File: docs/resources/helm_release.md [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+Description: |-
+    Creating the Helm Release resource.
+---
+
+# Helm Release
+
+The `tanzu-mission-control_helm_release` resource allows you to install, update, get and delete helm chart to a particular scope through Tanzu Mission Control.
+
+Before creating helm charts user needs to enable the helm service on a particular scope cluster or cluster group and for enable user can use `tanzu-mission-control_helm_feature` resource.
+there is one field `feature_ref` in `tanzu-mission-control_helm_release` which user can use for clean up purpose of the resources when they wanted to disable the hepl service.
```
Comment (ishangupta-ds) [inline docs/resources/helm_release.md:12]:
```suggestion
The`feature_ref` field of `tanzu-mission-control_helm_release` when specified, ensures clean up of this Terraform resource from the state file by creating a dependency on the Helm feature when the Helm feature is disabled.
```

File: internal/resources/helmfeature/resource_mock_test.go [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+	HelmFeatureEndpoint := (helper.ConstructRequestURL(https, endpoint, clAPIVersionAndGroup, testConfig.ScopeHelperResources.Cluster.Name, helmAPIKind)).String()
+
+	httpmock.RegisterResponder("POST", HelmFeatureEndpoint,
+		bodyInspectingResponder(t, postHelmFeatureRequest, 200, postHelmFeatureResponse))
+
+	httpmock.RegisterResponder("GET", HelmFeatureEndpoint,
+		bodyInspectingResponder(t, nil, 200, listResponse))
+
+	// httpmock.RegisterResponder("DELETE", HelmFeatureEndpoint,
+	// 	bodyInspectingResponder(t, nil, 200, nil))
```
Comment (ishangupta-ds) [inline internal/resources/helmfeature/resource_mock_test.go:148]: Can you remove the commented blocks if not required?

File: internal/resources/helmrelease/resource_mock_test.go [hunk 6]
```diff
@@ -0,0 +1,762 @@
...
+		bodyInspectingResponder(t, postRequest, 200, postResponse))
+
+	httpmock.RegisterResponder("PUT", getHelmReleaseEndpoint,
+		bodyInspectingResponder(t, putRequest, 200, putResponse))
+
+	httpmock.RegisterResponder("GET", getHelmReleaseEndpoint,
+		bodyInspectingResponder(t, nil, 200, getResponse))
+
+	httpmock.RegisterResponder("DELETE", deleteEndpoint, changeStateResponder(
+		// Set up the get to return 404 after the Secret has been 'deleted'.
```
Comment (ishangupta-ds) [inline internal/resources/helmrelease/resource_mock_test.go:306]: Secret?

File: internal/resources/helmrelease/resource_mock_test.go [hunk 7]
```diff
@@ -0,0 +1,762 @@
...
+		bodyInspectingResponder(t, postCGRequest, 200, postCGResponse))
+
+	httpmock.RegisterResponder("PUT", getCGHelmReleaseEndpoint,
+		bodyInspectingResponder(t, putCGRequest, 200, putCGResponse))
+
+	httpmock.RegisterResponder("GET", getCGHelmReleaseEndpoint,
+		bodyInspectingResponder(t, nil, 200, getCGResponse))
+
+	httpmock.RegisterResponder("DELETE", deleteCGEndpoint, changeStateResponder(
+		// Set up the get to return 404 after the Secret has been 'deleted'.
```
Comment (ishangupta-ds) [inline internal/resources/helmrelease/resource_mock_test.go:387]: Please update all such occurences.

File: internal/resources/helmrelease/resourec_helm_release.go [hunk 8]
```diff
@@ -66,7 +67,7 @@ func getHelmReleaseSchema(isDataSource bool) map[string]*schema.Schema {
 		featureRefKey: {
 			Type:        schema.TypeString,
 			Description: "Name of Namespace.",
```
Comment (ishangupta-ds) [inline internal/resources/helmrelease/resourec_helm_release.go:69]: Why is it namespace?

File: internal/resources/helmrelease/resourec_helm_release.go [hunk 9]
```diff
@@ -371,26 +372,49 @@ func updateCheckForSpec(d *schema.ResourceData, atomicSpec *releaseclustermodel.
...
+		if len(resp.Helms) == 0 {
+			return errors.Errorf("Tanzu mission control helm feature is disable on cluster group, name: %s", scopedFullnameData.FullnameCluster.ClusterName)
+		}
+
+		if resp.Helms[0].Status.Phase == nil {
+			return errors.Errorf("Tanzu mission control helm feature is disable on cluster group, name: %s", scopedFullnameData.FullnameCluster.ClusterName)
+		}
 	}
 
 	return nil
```
Comment (ishangupta-ds) [inline internal/resources/helmrelease/resourec_helm_release.go:420]: Please fix the file name. Verify other filenames as well.

File: templates/resources/helm_release.md.tmpl [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+Description: |-
+    Creating the Helm Release resource.
+---
+
+# Helm Release
+
+The `tanzu-mission-control_helm_release` resource allows you to install, update, get and delete helm chart to a particular scope through Tanzu Mission Control.
+
+Before creating helm charts user needs to enable the helm service on a particular scope cluster or cluster group and for enable user can use `tanzu-mission-control_helm_feature` resource.
+there is one field `feature_ref` in `tanzu-mission-control_helm_release` which user can use for clean up purpose of the resources when they wanted to disable the hepl service.
```
Comment (ishangupta-ds) [inline templates/resources/helm_release.md.tmpl:12]: Please update as per the above.

**Review by ishangupta-ds — CHANGES_REQUESTED** [review]

File: examples/resources/helmfeature/cl_resource.tf (hunk 1 above)
Comment (Shubbhang351) [reply examples/resources/helmfeature/cl_resource.tf]: Updated PTAL

File: internal/resources/helmfeature/resource_helm_feature_test.go (hunk 2 above)
Comment (Shubbhang351) [reply internal/resources/helmfeature/resource_helm_feature_test.go:60]: Updated PTAL

File: internal/resources/helmrelease/test_helper.go (hunk 3 above)
Comment (Shubbhang351) [reply internal/resources/helmrelease/test_helper.go]: Updated PTAL

File: docs/resources/helm_release.md (hunk 4 above)
Comment (Shubbhang351) [reply docs/resources/helm_release.md]: Updated PTAL

File: internal/resources/helmfeature/resource_mock_test.go (hunk 5 above)
Comment (Shubbhang351) [reply internal/resources/helmfeature/resource_mock_test.go]: Updated PTAL

File: internal/resources/helmrelease/resource_mock_test.go (hunk 6 above)
Comment (Shubbhang351) [reply internal/resources/helmrelease/resource_mock_test.go]: Updated PTAL

File: internal/resources/helmrelease/resource_mock_test.go (hunk 7 above)
Comment (Shubbhang351) [reply internal/resources/helmrelease/resource_mock_test.go]: Updated PTAL

File: internal/resources/helmrelease/resourec_helm_release.go (hunk 8 above)
Comment (Shubbhang351) [reply internal/resources/helmrelease/resourec_helm_release.go]: Updated PTAL

File: internal/resources/helmrelease/resourec_helm_release.go (hunk 9 above)
Comment (Shubbhang351) [reply internal/resources/helmrelease/resourec_helm_release.go:420]: Updated PTAL

File: templates/resources/helm_release.md.tmpl (hunk 10 above)
Comment (Shubbhang351) [reply templates/resources/helm_release.md.tmpl]: Updated PTAL

File: docs/resources/helm_feature.md [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+- `scope` (Block List, Min: 1, Max: 1) Scope for the Helm feature, having one of the valid scopes: cluster, cluster_group. (see [below for nested schema](#nestedblock--scope))
+
+### Optional
+
+- `meta` (Block List, Max: 1) Metadata for the resource (see [below for nested schema](#nestedblock--meta))
+
+### Read-Only
+
+- `id` (String) The ID of this resource.
+- `status` (Map of String) Status for the Repository.
```
Comment (ishangupta-ds) [inline docs/resources/helm_feature.md:80]: Should it be phase?

**Review by ishangupta-ds — CHANGES_REQUESTED** [review]

**Review by ramya-bangera — APPROVED** [review]

File: docs/resources/helm_release.md [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+<a id="nestedblock--spec"></a>
+### Nested Schema for `spec`
+
+Required:
+
+- `chart_ref` (Block List, Min: 1, Max: 1) Reference to the chart which will be installed. (see [below for nested schema](#nestedblock--spec--chart_ref))
+
+Optional:
+
+- `inline_config` (String) Inline values in yaml format.
```
[the remark this answers is not in the document]
Comment (Shubbhang351) [reply docs/resources/helm_release.md]: Updated PTAL

File: docs/resources/helm_release.md [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+- `interval` (String) Interval at which to reconcile the Helm release. This is the interval at which Tanzu Mission Control will attempt to reconcile changes in the helm release to the cluster. A sync interval of 0 would result in no future syncs. If no value is entered, a default interval of 5 minutes will be applied as `5m`.
+- `target_namespace` (String) Name of target namespace.
```
[the remark this answers is not in the document]
Comment (Shubbhang351) [reply docs/resources/helm_release.md]: Updated PTAL

File: docs/resources/helm_release.md [hunk 14]
```diff
@@ -0,0 +1,289 @@
...
+- `git_repository` (Block List, Max: 1) Git repository type. (see [below for nested schema](#nestedblock--spec--chart_ref--git_repository))
+- `helm_repository` (Block List, Max: 1) Helm repository type. (see [below for nested schema](#nestedblock--spec--chart_ref--helm_repository))
+
+<a id="nestedblock--spec--chart_ref--git_repository"></a>
+### Nested Schema for `spec.chart_ref.git_repository`
+
+Required:
+
+- `chart_path` (String) Path of the chart in the git repository.
+- `repository_name` (String) Repository name.
```
[the remark this answers is not in the document]
Comment (Shubbhang351) [reply docs/resources/helm_release.md]: Updated PTAL

File: docs/resources/helm_release.md [hunk 15]
```diff
@@ -0,0 +1,289 @@
...
+- `repository_namespace` (String) Repository namespace.
```
[the remark this answers is not in the document]
Comment (Shubbhang351) [reply docs/resources/helm_release.md]: no It's not the branch

File: docs/resources/helm_release.md [hunk 16]
```diff
@@ -0,0 +1,289 @@
...
+
+
+<a id="nestedblock--spec--chart_ref--helm_repository"></a>
+### Nested Schema for `spec.chart_ref.helm_repository`
+
+Required:
+
+- `chart_name` (String) Name of the chart in the helm repository.
+- `repository_name` (String) Repository name.
+- `repository_namespace` (String) Repository namespace.
```
[the remark this answers is not in the document]
Comment (Shubbhang351) [reply docs/resources/helm_release.md]: Updated PTAL

File: docs/resources/helm_release.md [hunk 17]
```diff
@@ -0,0 +1,289 @@
...
+- `phase` (String)
+
+<a id="nestedobjatt--status--generated_resources"></a>
+### Nested Schema for `status.generated_resources`
+
+Read-Only:
+
+- `cluster_role_name` (String)
+- `role_binding_name` (String)
+- `service_account_name` (String)
```
[the remark this answers is not in the document]
Comment (Shubbhang351) [reply docs/resources/helm_release.md]: Updated PTAL

File: docs/resources/helm_feature.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (Shubbhang351) [reply docs/resources/helm_feature.md]: Updated PTAL

File: docs/resources/helm_release.md [hunk 18]
```diff
@@ -0,0 +1,289 @@
...
+- `uid` (String) UID of the resource
+
+
+<a id="nestedatt--status"></a>
+### Nested Schema for `status`
+
+Read-Only:
+
+- `generated_resources` (List of Object) (see [below for nested schema](#nestedobjatt--status--generated_resources))
+- `phase` (String)
```
[the remark this answers is not in the document]
Comment (Shubbhang351) [reply docs/resources/helm_release.md]: Updated PTAL

File: docs/resources/helm_feature.md (hunk 11 above)
Comment (Shubbhang351) [reply docs/resources/helm_feature.md]: it's at the Upper level of the schema, so it should be status.

**Review by ishangupta-ds — APPROVED** [review]
LGTM!

**Pull request merged**